File: Graph/547. Number of Provinces/547. Number of Provinces.py
#problem link: https://leetcode.com/problems/number-of-provinces

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        provinces = 0
        visited = [False for i in range(n)]

        def dfs(ind: int, isConnected: List[List[int]], visited: List[bool]): 
            visited[ind] = True

            for j in range(0, len(isConnected)):
                if ind != j and isConnected[ind][j] == 1:
                    if visited[j] == False:
                        dfs(j, isConnected, visited)

        for i in range(0, n):
            if visited[i] == False:
                provinces += 1
                dfs(i, isConnected, visited)
        
        return provinces

File: Graph/547. Number of Provinces/547. Number of Provinces.cpp
// problem link: https://leetcode.com/problems/number-of-provinces/

#include<iostream>
#include<vector>
using namespace std;

class Solution {
private:
    void dfs(int ind, vector<vector<int>>& isConnected, vector<int>& visited) {
        visited[ind] = 1;
        for (int j = 0; j < isConnected.size(); j++) {
            if (ind != j && isConnected[ind][j] == 1) {
                if (visited[j] == false) {
                    dfs(j, isConnected, visited);
                }
            }
        }
    }
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int provinces = 0;
        int n = isConnected.size();
        vector<int> visited(n, 0);
        
        for (int i = 0; i < n; i++) {
            if (visited[i] == 0) {
                provinces++;
                dfs(i, isConnected, visited);
            }
        }

        return provinces;
    }
};